allennlp/data/dataset_readers/reading_comprehension/triviaqa.py

```diff
@@ -1,32 +1,245 @@
+# pylint: disable=no-member
 import json
 import logging
-import os
+import pathlib
+import shutil
 import tarfile
-from typing import Dict, List, Tuple
+import tempfile
+from typing import Dict, List, Tuple, Iterator, Iterable, NamedTuple
 
 from overrides import overrides
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.metrics import pairwise_distances
+import numpy as np
 
-from allennlp.common import Params
-from allennlp.common.file_utils import cached_path
+from allennlp.common import Params, JsonDict
+from allennlp.common.checks import ConfigurationError
+from allennlp.common.util import lazy_groups_of
 from allennlp.data.dataset_readers.dataset_reader import DatasetReader
 from allennlp.data.instance import Instance
 from allennlp.data.dataset_readers.reading_comprehension import util
 from allennlp.data.token_indexers import SingleIdTokenIndexer, TokenIndexer
 from allennlp.data.tokenizers import Token, Tokenizer, WordTokenizer
+from allennlp.data.tokenizers.token import json_to_token, token_to_json, truncate_token
 
 logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
 
+_PARAGRAPH_TOKEN = "@@PARAGRAPH@@"
+
+# We create NamedTuples for storing "merged paragraphs" and questions
+# to make it easier to preprocess the dataset into JSONL format
+# and start training from the processed data.
+
+class MergedParagraphs(NamedTuple):
+    """
+    A typical TriviaQA question has multiple files with answers, each consisting
+    of many small paragraphs. We merge these into fewer paragraphs of larger size.
+    """
+    texts: List[str]
+    tokens: List[List[Token]]
+    token_spans: List[List[Tuple[int, int]]]
+    has_answers: List[int] = None
```

Comment:
Why is this an `int`?

Reply:
It is a list of indexes for merged paragraphs which contain the answer.
For example if we have 5 paragraphs and the first 2 paragraphs contain the answer then the list will be [0,1].

We could get rid of this variable and use the token_spans list to determine which paragraphs have answers. Not sure if you prefer that?